quintilesims/layer0-log-forwarder
# Module implementation

First commit - adds functionality for the terraform module including documentation (README.md) and development workflow (DEVELOP.md).

**How to test it**
1. Ensure you have a layer0 instance by creating a new one or reusing an existing one
2. Navigate to the `examples` directory and populate the variables either via a `tfvars` file like the one below:
```
endpoint = "https://<l0-endpoint>"
token = "<l0-token>"
access_key = "abc"
secret_key = "def"
region = "us-west-2"
kinesis_stream_name = "salesforce_splunk"
stream_aws_access_key = "<dev-ops-access-key>"
stream_aws_secret_key = "<dev-ops-secret-key>"
stream_aws_region = "us-west-2"
layer0_prefix = "<l0-prefix>"
``` 
3. Apply the terraform file
4. Create another service in the same layer0 instance which generates cloudwatch logs. You can use something like our guestbook sample which creates log entries for every entry you add to the guestbook.
5. Navigate to our splunk instance and query `host=layer0`. 
6. Confirm you can see the logs from your service in Splunk.

**Note:**
The example directory has a sample terraform configuration for how to use this module. Note however that the module points to a github repo which doesn't exist yet as this PR is intended to add the module to this repo.

So just for the test, in `examples/main.tf`

Instead of:
```
module "layer0_log_forwarder" {
  source                = "github.com/quintiles/layer0-log-forwarder/terraform"
  ...
}
```

Update it to a relative path:
```
module "layer0_log_forwarder" {
  source                = "./../terraform"
  ...
}
```

Also for the stream name - you can use `salesforce_splunk` in the devops account as no one is writing to that stream currently.

**Caveats**
There seems to be an issue with the aws terraform resrouce: `aws_cloudwatch_log_subscription_filter`. The state of this seems to be eventually consistent, which in my observations leads to creates, deletes and recreations to fail.
